### ISSUE TITLE
Macro typos generating exponentially growing token streams cause rustc to take exponentially long

### ISSUE THREAD

**Issue by CAD97** [rust-lang/rust#97747]
I tried this code:

```rust
#![recursion_limit="32"]
macro_rules! m { ($($args:tt)*) => { m! { $($args)* $($args)* } } }
m! { test }
```

**I expected to see this happen:** compilation would terminate after a reasonable amount of time.

**Instead, this happened:** process aborted with a stack overflow after six minutes of pegging a single core.

([The actual code which ran into this issue.](https://twitter.com/CAD97_/status/1532913123006590977?s=20&t=1DbdSrb0SVclkQz5NAHiBg) Above is a minimal reproducer.)

I did some additional timing. On my AMD Ryzen 7 3700X 8-Core Processor, I got the following timings:

recursion limit | time to compilation failure
------- | -------
24 | 4 seconds
25 | 8 seconds
26 | 16 seconds
27 | 32 seconds

The suggested mitigation would be to keep track of token lengths of streams to catch degenerate macros like the above. Ideally this should track the *deep* length, but tracking shallow length (not including the length of nested token trees) should still catch unbounded growth like this, just a bit slower. If we assume the average token length is eight bytes, 2<sup>27</sup> tokens is a GiB of expanded code.

### Meta

`rustc --version --verbose`:
```
rustc 1.63.0-nightly (a6b8c6954 2022-06-03)
binary: rustc
commit-hash: a6b8c6954829669a5c4fa320c3e6132edf04fcfc
commit-date: 2022-06-03
host: x86_64-pc-windows-msvc
release: 1.63.0-nightly
LLVM version: 14.0.4
```
